File: 2024/day02.go
package main

import (
	"fmt"
)

import _ "embed"

import "aoc/tools"

var DEBUG bool = false
var TEST bool = false

var test = `
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9`[1:]

//go:embed day02.txt
var live string

func is_safe(row []int) bool {
	for i := 0; i < len(row)-1; i++ {
		x := row[i]
		y := row[i+1]
		if (y-x)*(row[1]-row[0]) < 0 {
			return false
		}
		if tools.AbsInt(y-x) < 1 || tools.AbsInt(y-x) > 3 {
			return false
		}
	}
	return true
}

func part1(data [][]int) int {
	sumx := 0
	for _, row := range data {
		if is_safe(row) {
			sumx++
		}
	}
	return sumx
}

func part2(data [][]int) int {
	safe := 0
	for _, row := range data {
		if is_safe(row) {
			safe++
		} else {
			for i := 0; i < len(row); i++ {
				if is_safe(tools.Remove(row, i)) {
					safe++
					break
				}
			}
		}
	}
	return safe
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)

	data := tools.Parse[int](input)

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}

File: 2017/day02.py
import sys
TEST = 'test' in sys.argv
DEBUG = 'debug' in sys.argv

test = """\
5 1 9 5
7 5 3
2 4 6 8"""

test2 = """\
5 9 2 8
9 4 7 3
3 8 6 5"""

live = open('day2.txt').read()

def parse(lines):
    data = []
    for line in lines.splitlines():
        data.append( list(map(int, line.split())))
    return data

data = test if TEST else live
data = parse(data)

def part1(data):
    return sum(max(row)-min(row) for row in data)

def part2(data):
    sum = 0
    for row in data:
        for i,a in enumerate(row):
            for b in row[i+1:]:
                a1 = max(a,b)
                b1 = min(a,b)
                if a1 % b1 == 0:
                    sum += a1//b1
                    break
    return sum

print('Part 1:', part1(data))
if TEST:
    data = parse(test2)
print('Part 2:', part2(data))
